DS-4021: Updated patch operations for Item repository and added patch endpoints for EPerson.

mspalti opened the pull request:
This PR includes new patch operation support classes and new patch endpoints for EPerson operations.  The approach is modeled on existing submit patch operations and should be reusable for other DSpaceObject types.

- Uses factory method to return autowired patch operations.
- Currently assumes separate factory for each DSpaceObject (Collections, EPerson, etc.)
- Implements Item withdrawn and discoverable patch operations using the new patch classes.
- Implements new EPerson operations.
- Includes support for LateObjectEvaluator methods via the PatchOperation (these are not used for withdrawn and discoverable operations but were tested with a dummy class and MetadataValueRest[]).
- Includes modifications to the rest contract (Rest7Contract will need to be updated if this PR is accepted).
- Includes integration tests.

mspalti committed:
Updated respository patch classes and ItemRestRepository.

tomdesair commented:
Hi @mspalti, thanks for the contribution!

Could you please add integration tests in class `ItemRestRepositoryIT` class for the new patch operations on the `discoverable` and `withdrawn` fields? Maybe I missed them because there are a lot of identation changes in that class. You should find similar tests for other patch operations in that class.

Thanks!

mspalti committed:
Rolling back indentation changes in ItemRestRepositoryIT.

Reverting indentation changes in ItemRestRepositoryIT.

More updates to indentation.

Corrected line length errors in ItemRestRepositoryIT.

Final formatting updates for ItemRestRepositoryIT.

mspalti commented:
Hi @tomdesair 

Sorry about the indentation changes in `ItemRestRepositoryIT`. I removed indentation changes so the diff is cleaner now. The tests that you requested are there from previous work that I did on `ItemRestRepository.`

tdonohue commented:
@mspalti : Could you link this new PR up to the JIRA ticket it refers to?  That helps reviewers to better understand the context (without having to go searching around).

mspalti commented:
Thanks for the heads up, @tdonohue. This PR relates to work done originally for DS-3909.  DS-3908 is also related.

mspalti committed:
[DS-4021] Added patch method for EPerson.

Implements patch operations for password, canLogin, requireCertificate and netid. Includes integration tests.

mspalti commented:
The community sprint 3 includes an issue [DS-4021](https://jira.duraspace.org/browse/DS-4021) that requires the work done for this PR.  I have  renamed this PR to reference the new Jira issue and have included work on new EPerson patch operations.

mspalti committed:
DS-4021 Reintroduce abstract and implementation classes for patch operations.

Prior to this commit, the factory implementations used only path and value, ignoring the operation name. This commit reintroduces default implementations operations (add, move, remove, replace) and working implementations for eperson and item replace operations. The factory methods are now designed to be used specifically for replace operations. As needed, new factory methods can be added.

mspalti commented:
This should probably be labelled "work in progress" (can I do that?)

I just added a commit, reintroducing `AbstractResourcePatch` and concrete implementations for `Item` and `EPerson` replace operations. The code in this PR was previously ignoring the operation name and using only the path and value.

abollini commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/EPersonRestRepositoryIT.java:
```diff
@@ -507,11 +512,226 @@ public void deleteViolatingConstraints() throws Exception {
+                        .andExpect(status().isBadRequest());
+
+
     }
```
We need more test to verify that:
- the single replace operation fails for anonymous or not administrator users
- multiple valid replace operations in the same requests succeed 
- multiple replace operations where the second to last is invalid will fails without any changes (including the one required by the valid operations)

mspalti replied:
If I understand the 3rd test correctly, l need to validate all operations before beginning to process them in sequence.  Otherwise the first operations will introduce changes before the later operation fails. I will start work on this (it seems like the correct behaviour!) but let me know if you are thinking of something else.

abollini replied:
your understanding is correct thanks!
From the json patch specification

>  Note that the HTTP PATCH method is atomic, as per [RFC5789].
   Therefore, the following patch would result in no changes being made
   to the document at all (because the "test" operation results in an
   error):
`
   [
     { "op": "replace", "path": "/a/b/c", "value": 42 },
     { "op": "test", "path": "/a/b/c", "value": "C" }
   ]
`

tdonohue commented:
@mspalti : I've labeled it "work in progress" on your behalf.  I think GitHub only lets project "owners" add labels at this time...so, since you have developer access, I don't know that you can add your own labels yet.

Once you want this label removed, just let us know or tell us this is "ready for a final review" (and we'll remove the label)

abollini commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/EPersonRestRepositoryIT.java:
```diff
@@ -507,11 +512,226 @@ public void deleteViolatingConstraints() throws Exception {
+                .content(patchBody)
+                .contentType(MediaType.APPLICATION_JSON_PATCH_JSON))
+                        .andExpect(status().isBadRequest());
+
```
it will be better to check that no changes occurred

abollini commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/ItemRestRepositoryIT.java:
```diff
@@ -89,11 +89,15 @@ public void findAllTest() throws Exception {
         getClient(token).perform(get("/api/core/items"))
                    .andExpect(status().isOk())
                    .andExpect(jsonPath("$._embedded.items", Matchers.containsInAnyOrder(
```
I'm not able to find any real changes other than code format in this file. Is any?

mspalti replied:
One small change.  The operation value is now a string ("true"), not a boolean.  

The reformatting in this file was due to a checkstyle error (exceeded 120 character limit). In some cases the reformatting was probably not needed (caused by erroneous indentation, sorry!).

abollini commented:
To limit access to the patch operation to the administrator you can use spring security see
https://github.com/DSpace/DSpace/blob/master/dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/EPersonRestRepository.java#L113

it could be useful to check how to allow to specific patch operation to apply different security constraints, but this could be not trivial see
https://github.com/DSpace/DSpace/pull/1976#issuecomment-415054136
and I suggest to leave it for a subsequent task

mspalti committed:
[DS-4021] This commit includes code formatting and comment changes.

Also a few minor modifications to patch operations.

Minor code format changes.

Final code formatting changes for EPersonRestRepositoryIT.

mspalti committed:
[DS-4021] Patch processing now operates on the rest model and commits changes after successful completion of all patch operations.

Also added new integration tests.

Minor changes to error handling and comments.

[DS-4021] Updated tests and javadocs.

Also updated the string comparison in EPersonRestRespository that triggers the update of the netid after a patch operation is complete.

mspalti commented:
I think we can remove the work-in-progress label. This code is ready for review.

The latest changes changes:

- Added @PreAuthorize annotation to restrict access to Admin
- Converted patch classes to work with the RestModel
- ItemRestRepository and EPsersonRestRepository methods now updating the business object and context after patch operations successfully complete.
- Added the tests requested by @abollini (tests for authorization and multiple operations)

tdonohue commented on dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/patch/AbstractResourcePatch.java:
```diff
@@ -26,85 +22,86 @@
-        remove(restModel, context, operation);
-        add(restModel, context, operation);
+       // remove(restModel, context, operation);
+       // add(restModel, context, operation);
```
As you commented these out, it looks like we should remove all 5 lines of comments here, as they seem inapplicable?

tdonohue commented on dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/patch/factories/impl/PatchOperation.java:
```diff
@@ -0,0 +1,71 @@
+
+    // This is duplicated code (see org.dspace.app.rest.submit.factory.impl.PatchOperation)
+    // If it stays here, it should be DRY. Current patch resource patch operations do not
+    // use these methods.
```
Were you able to determine if this code is needed here?  It seems like if it's not used, we shouldn't include it here until it *is* used.

mspalti replied:
I was going to ask about this one. I definitely agree with you. I'm anticipating a need.  

These methods evaluate values that are Json objects or arrays, but as I have implemented the current patch operations values are provided as Strings. I could of course force the issue (e.g. "value": { "netid": "newNetId" }) but obviously won't -- unless there's some actual reason to do this.

Still, I'm curious if there things on the current roadmap that will require the ability to type check a json object or array when patching a resource? This is on the current todo list then perhaps these should stay.

mspalti replied:
Since the json patch converter returns JsonValueEvaluator instances that require the type of evaluation provided by these methods, I think a complete implementation of PatchOperation requires them in principle. So at the risk of violating the YAGNI principle I think they should remain.

tdonohue commented on dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/patch/factories/impl/PatchOperation.java:
```diff
@@ -0,0 +1,71 @@
+        return results;
+    }
+
+    public T evaluateSingleObject(LateObjectEvaluator value) {
```
Could you add Javadocs here please, if this method is used/need.

mspalti replied:
Will add javadoc if method is needed!

mspalti replied:
Added javadoc. It's not showing up here in the diff but view file and it's there. ...?

tdonohue reviewed (commented):
@mspalti : Code is looking great overall. I added a few, minor comments inline (minor code cleanup, really).  Thanks for all the great Integration Tests here!

abollini commented on dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/EPersonRestRepository.java:
```diff
@@ -179,6 +187,59 @@ public EPersonRest findByEmail(@Parameter(value = "email", required = true) Stri
+            ePerson.setCanLogIn(ePersonRest.isCanLogIn());
+        }
+        if (ePersonRest.getNetid() != null) {
+            if (!ePersonRest.getNetid().equals(ePerson.getName())) {
```
this looks odd

mspalti replied:
I think this was my attempt to avoid NPE.  After checking, I decided to replace with `!Objects.equals(ePersonRest.getNetid(), ePerson.getNetid())`. 

Perhaps a bit less odd.

abollini commented on dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/patch/factories/impl/EPersonCertificateReplaceOperation.java:
```diff
@@ -0,0 +1,70 @@
+            throws PatchBadRequestException {
+
+        checkOperationValue((String) operation.getValue());
+        Boolean requireCert = BooleanUtils.toBooleanObject((String) operation.getValue());
```
we should support the right behaviors that is to send a boolean instead than a string. It is also convenient to be friendly with the client and accept string, see here https://github.com/DSpace/DSpace/pull/2187/commits/89ebd2c154515b8bc41ca6fbab199dfb097399e8

abollini commented on dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/patch/factories/impl/ItemDiscoverableReplaceOperation.java:
```diff
@@ -0,0 +1,71 @@
+    private ItemRest replace(ItemRest item, Operation operation) {
+
+        checkOperationValue((String) operation.getValue());
+        Boolean discoverable = BooleanUtils.toBooleanObject((String) operation.getValue());
```
we must should receive  real boolean values and accept also boolean "String"

abollini commented on dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/patch/factories/impl/ItemWithdrawReplaceOperation.java:
```diff
@@ -0,0 +1,99 @@
+            throws PatchBadRequestException, UnprocessableEntityException {
+
+        checkOperationValue((String) operation.getValue());
+        Boolean withdraw = BooleanUtils.toBooleanObject((String) operation.getValue());
```
again accept boolean value

abollini commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/EPersonRestRepositoryIT.java:
```diff
@@ -514,4 +519,420 @@ public void deleteViolatingConstraints() throws Exception {
+        getClient(token).perform(patch("/api/eperson/epersons/" + ePerson.getID())
+                .content(patchBody)
+                .contentType(MediaType.APPLICATION_JSON_PATCH_JSON))
+                   .andExpect(status().isForbidden());
```
it will be better to check that no changes have been applied

tdonohue replied:
I agree with @abollini here. We should also add a check to see that no changes were made to the object... checking the return status is good, but we also want to ensure that the patch wasn't accidentally applied.  

Same with the other comment just below this one.

mspalti replied:
I think we are good for the missing tests to ensure that the patch was not applied. I added them (they don't appear here in the diff summary.)

This one checks the netid to be sure it wasn't updated.    
  ```
  getClient(token).perform(patch("/api/eperson/epersons/" + ePerson.getID())
                .content(patchBody)
                .contentType(MediaType.APPLICATION_JSON_PATCH_JSON))
                   .andExpect(status().isForbidden());


        token = getAuthToken(admin.getEmail(), password);


        getClient(token).perform(get("/api/eperson/epersons/" + ePerson.getID()))
                        .andExpect(status().isOk())
                        .andExpect(jsonPath("$.netid", Matchers.is(nullValue())));
```

abollini commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/EPersonRestRepositoryIT.java:
```diff
@@ -514,4 +519,420 @@ public void deleteViolatingConstraints() throws Exception {
+                .content(patchBody)
+                .contentType(MediaType.APPLICATION_JSON_PATCH_JSON))
+                        .andExpect(status().isUnauthorized());
+
```
as above, check for no changes

mspalti replied:
Done.

abollini commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/EPersonRestRepositoryIT.java:
```diff
@@ -514,4 +519,420 @@ public void deleteViolatingConstraints() throws Exception {
+                .contentType(MediaType.APPLICATION_JSON_PATCH_JSON))
+                   .andExpect(status().isOk())
+                   .andExpect(jsonPath("$.netid", Matchers.is("newNetId")));
+
```
I suggest to check that metadata outside the path operation stay untouched

abollini reviewed (changes requested):
this is a very good job, just some final remark. Thanks!

mspalti committed:
[DS-4021] Replace patch operations on a non-existent value now result in a PatchBadRequestException.

Added missing license headers.

mspalti committed:
[DS-4021] Replace patch operations on a non-existent value now result in a PatchBadRequestException.

Added missing license headers.

Added FIXME.

mspalti commented:
I apologize for the latest commit (I really thought this was done!).  All patch operations (except password for reasons explained in a TODO) now check to be sure they operate on an existing value, per requirements in the Rest7Contract.  Previously some tests were updating null values, and should have failed.

I added a method to the EPersonBuilder class so that the netid can be set during tests. Otherwise, none of the changes here will be surprising.

tdonohue commented on dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/patch/AbstractResourcePatch.java:
```diff
@@ -26,85 +22,81 @@
+    protected R replace(R restModel, Operation operation)
+            throws UnprocessableEntityException, PatchBadRequestException {
+        throw new UnprocessableEntityException(
+                "The remove operation is not supported."
```
Really tiny, picky thing. This should say `The replace operation is not supported.`

tdonohue commented on dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/patch/factories/impl/EPersonPasswordReplaceOperation.java:
```diff
@@ -0,0 +1,63 @@
+        checkOperationValue(operation.getValue());
+        if (eperson.getPassword() == null) {
+            /*
+             * FIXME: the password field in eperson rest model is always null because
```
Thanks for this detailed `FIXME`.  I'm wondering if we have (or should have) a bug ticket to describe how to fix this?

tdonohue reviewed (commented):
@mspalti : Gave this another review today. Overall, this is looking great.  I just had a few minor inline comments, but this is looking nearly ready-to-merge to me.

mspalti committed:
[DS-4021] Added a bit of IOC that helps to assure consistent checks before applying patch replace operations.

mspalti commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/EPersonRestRepositoryIT.java:
```diff
@@ -507,11 +512,226 @@ public void deleteViolatingConstraints() throws Exception {
-                .andExpect(status().isOk());
+                        .andExpect(status().isOk());
+    }
+
```
Done.

mspalti commented:
I believe all change requests are complete.  Happy to fine tune more tests if needed!

In the final commit I inverted things between the abstract replace operation class and the implementations; i.e. moved the `process` method into the parent class to assure that checks are applied consistently.  The change in flow control doesn't affect anything else.

tdonohue reviewed (approved):
👍 This looks good to me now. I think it's ready to merge.  However, I'd appreciate it if @abollini could give one last code review -- as he had several previous requests (which now look to be corrected).

abollini commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/EPersonRestRepositoryIT.java:
```diff
@@ -514,4 +520,480 @@ public void deleteViolatingConstraints() throws Exception {
+                                        .build();
+
+        List<Operation> ops = new ArrayList<Operation>();
+        ReplaceOperation replaceOperation = new ReplaceOperation("/canLogin", "true");
```
the right way to patch a boolean attribute is to send a real boolean instead than a String.
I suggest to make the implementation client friendly here and accept both literal boolean than "textual" boolean.
See here for an example of such implementation in the submission patches
https://github.com/DSpace/DSpace/commit/89ebd2c154515b8bc41ca6fbab199dfb097399e8

You could also add an additional test to check that textual boolean are still accepted and change this test to check the right boolean patch

abollini commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/EPersonRestRepositoryIT.java:
```diff
@@ -514,4 +520,480 @@ public void deleteViolatingConstraints() throws Exception {
+        String token = getAuthToken(admin.getEmail(), password);
+
+        List<Operation> ops = new ArrayList<Operation>();
+        ReplaceOperation replaceOperation = new ReplaceOperation("/canLogin", "true");
```
same as above, we need to support real boolean and be friendly with textual boolean

abollini commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/EPersonRestRepositoryIT.java:
```diff
@@ -514,4 +520,480 @@ public void deleteViolatingConstraints() throws Exception {
+
+        List<Operation> ops = new ArrayList<Operation>();
+
+        ReplaceOperation replaceOperation1 = new ReplaceOperation("/canLogin", "true");
```
better to use the official way here true instead than "true"

abollini commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/EPersonRestRepositoryIT.java:
```diff
@@ -514,4 +520,480 @@ public void deleteViolatingConstraints() throws Exception {
+        ops.add(replaceOperation1);
+        ReplaceOperation replaceOperation2 = new ReplaceOperation("/netid", "multitestId");
+        ops.add(replaceOperation2);
+        ReplaceOperation replaceOperation3 = new ReplaceOperation("/certificate", "true");
```
better to use the official way here true instead than "true"

abollini commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/EPersonRestRepositoryIT.java:
```diff
@@ -514,4 +520,480 @@ public void deleteViolatingConstraints() throws Exception {
+        String token = getAuthToken(admin.getEmail(), password);
+
+        List<Operation> ops = new ArrayList<Operation>();
+        ReplaceOperation replaceOperation0 = new ReplaceOperation("/canLogin", "true");
```
better to use the official way here true instead than "true"

abollini commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/EPersonRestRepositoryIT.java:
```diff
@@ -514,4 +520,480 @@ public void deleteViolatingConstraints() throws Exception {
+        // The first operation in the series sets canLogIn to be false.
+        ReplaceOperation replaceOperation1 = new ReplaceOperation("/canLogin", "false");
+        ops.add(replaceOperation1);
+        ReplaceOperation replaceOperation2 = new ReplaceOperation("/certificate", "true");
```
better to use the official way here true instead than "true"

abollini commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/EPersonRestRepositoryIT.java:
```diff
@@ -514,4 +520,480 @@ public void deleteViolatingConstraints() throws Exception {
+                        .andExpect(jsonPath("$.canLogIn", Matchers.is(true)));
+
+        // The first operation in the series sets canLogIn to be false.
+        ReplaceOperation replaceOperation1 = new ReplaceOperation("/canLogin", "false");
```
better to use the official way here true instead than "true"

abollini commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/EPersonRestRepositoryIT.java:
```diff
@@ -514,4 +520,480 @@ public void deleteViolatingConstraints() throws Exception {
+        // This will fail.
+        ReplaceOperation replaceOperation3 = new ReplaceOperation("/cert", "false");
+        ops.add(replaceOperation3);
+        ReplaceOperation replaceOperation4 = new ReplaceOperation("/certificate", "false");
```
better to use the official way here true instead than "true"

abollini commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/EPersonRestRepositoryIT.java:
```diff
@@ -514,4 +520,480 @@ public void deleteViolatingConstraints() throws Exception {
+        ReplaceOperation replaceOperation2 = new ReplaceOperation("/certificate", "true");
+        ops.add(replaceOperation2);
+        // This will fail.
+        ReplaceOperation replaceOperation3 = new ReplaceOperation("/cert", "false");
```
it is not really necessary to use a real boolean here as the path doesn't exist but... 
Maybe it is better to clarify that the operation will fail due to the missing path (or you can use "/unexistentPath", "someValue") to avoid confusion. Someone could assume later that the issue was the textual boolean

abollini commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/ItemRestRepositoryIT.java:
```diff
@@ -356,7 +376,7 @@ public void withdrawPatchTest() throws Exception {
 
         List<Operation> ops = new ArrayList<Operation>();
-        ReplaceOperation replaceOperation = new ReplaceOperation("/withdrawn", true);
+        ReplaceOperation replaceOperation = new ReplaceOperation("/withdrawn", "true");
```
better to use the official way here true instead than "true"

abollini commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/ItemRestRepositoryIT.java:
```diff
@@ -410,7 +430,7 @@ public void withdrawPatchUnauthorizedTest() throws Exception {
 
         List<Operation> ops = new ArrayList<Operation>();
-        ReplaceOperation replaceOperation = new ReplaceOperation("/withdrawn", true);
+        ReplaceOperation replaceOperation = new ReplaceOperation("/withdrawn", "true");
```
better to use the official way here true instead than "true"

abollini commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/ItemRestRepositoryIT.java:
```diff
@@ -458,7 +478,7 @@ public void withdrawPatchForbiddenTest() throws Exception {
 
         List<Operation> ops = new ArrayList<Operation>();
-        ReplaceOperation replaceOperation = new ReplaceOperation("/withdrawn", true);
+        ReplaceOperation replaceOperation = new ReplaceOperation("/withdrawn", "true");
```
better to use the official way here true instead than "true"

abollini commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/ItemRestRepositoryIT.java:
```diff
@@ -578,7 +598,7 @@ public void reinstatePatchTest() throws Exception {
 
         List<Operation> ops = new ArrayList<Operation>();
-        ReplaceOperation replaceOperation = new ReplaceOperation("/withdrawn", false);
+        ReplaceOperation replaceOperation = new ReplaceOperation("/withdrawn", "false");
```
better to use the official way here true instead than "true"

abollini commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/ItemRestRepositoryIT.java:
```diff
@@ -637,7 +657,7 @@ public void reinstatePatchUnauthorizedTest() throws Exception {
 
         List<Operation> ops = new ArrayList<Operation>();
-        ReplaceOperation replaceOperation = new ReplaceOperation("/withdrawn", false);
+        ReplaceOperation replaceOperation = new ReplaceOperation("/withdrawn", "false");
```
better to use the official way here true instead than "true"

abollini commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/ItemRestRepositoryIT.java:
```diff
@@ -686,15 +706,15 @@ public void reinstatePatchForbiddenTest() throws Exception {
 
         List<Operation> ops = new ArrayList<Operation>();
-        ReplaceOperation replaceOperation = new ReplaceOperation("/withdrawn", false);
+        ReplaceOperation replaceOperation = new ReplaceOperation("/withdrawn", "false");
```
better to use the official way here true instead than "true"

abollini commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/ItemRestRepositoryIT.java:
```diff
@@ -730,7 +750,7 @@ public void makeDiscoverablePatchTest() throws Exception {
 
         List<Operation> ops = new ArrayList<Operation>();
-        ReplaceOperation replaceOperation = new ReplaceOperation("/discoverable", true);
+        ReplaceOperation replaceOperation = new ReplaceOperation("/discoverable", "true");
```
better to use the official way here true instead than "true"

abollini commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/ItemRestRepositoryIT.java:
```diff
@@ -783,7 +803,7 @@ public void makeDiscoverablePatchUnauthorizedTest() throws Exception {
 
         List<Operation> ops = new ArrayList<Operation>();
-        ReplaceOperation replaceOperation = new ReplaceOperation("/discoverable", true);
+        ReplaceOperation replaceOperation = new ReplaceOperation("/discoverable", "true");
```
better to use the official way here true instead than "true"

abollini commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/ItemRestRepositoryIT.java:
```diff
@@ -827,7 +847,7 @@ public void makeDiscoverablePatchForbiddenTest() throws Exception {
 
         List<Operation> ops = new ArrayList<Operation>();
-        ReplaceOperation replaceOperation = new ReplaceOperation("/discoverable", true);
+        ReplaceOperation replaceOperation = new ReplaceOperation("/discoverable", "true");
```
better to use the official way here true instead than "true"

abollini commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/ItemRestRepositoryIT.java:
```diff
@@ -869,17 +889,17 @@ public void makeUnDiscoverablePatchTest() throws Exception {
 
         List<Operation> ops = new ArrayList<Operation>();
-        ReplaceOperation replaceOperation = new ReplaceOperation("/discoverable", false);
+        ReplaceOperation replaceOperation = new ReplaceOperation("/discoverable", "false");
```
better to use the official way here true instead than "true"

abollini commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/ItemRestRepositoryIT.java:
```diff
@@ -914,15 +934,15 @@ public void makeUnDiscoverablePatchUnauthorizedTest() throws Exception {
 
         List<Operation> ops = new ArrayList<Operation>();
-        ReplaceOperation replaceOperation = new ReplaceOperation("/discoverable", false);
+        ReplaceOperation replaceOperation = new ReplaceOperation("/discoverable", "false");
```
better to use the official way here true instead than "true"

abollini commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/ItemRestRepositoryIT.java:
```diff
@@ -958,15 +978,15 @@ public void makeUnDiscoverablePatchForbiddenTest() throws Exception {
 
         List<Operation> ops = new ArrayList<Operation>();
-        ReplaceOperation replaceOperation = new ReplaceOperation("/discoverable", false);
+        ReplaceOperation replaceOperation = new ReplaceOperation("/discoverable", "false");
```
better to use the official way here true instead than "true"

abollini reviewed (approved):
Can I ask for some additional minor improvements here? all are related to the use of real boolean instead than textual boolean.
I'm already approving the PR as it is also fine for me to merge it as is and open a new ticket for these further improvements if @mspalti has no time to do it right now. I really like to have this PR merged soon

mspalti committed:
[DS-4021] Updated integration tests.

Merge branch 'repositorypatch' of https://github.com/mspalti/DSpace into repositorypatch

mspalti committed:
[DS-4021] Corrected minor typo in integration test.

mspalti commented:
I think the integration tests now have the requested changes.  I changed to boolean values for existing tests, and added a tests to check for string to boolean conversion.

abollini commented:
@mspalti thanks for all the hard work here!
